pkg/odo/cli/component/deploy.go

```diff
@@ -0,0 +1,221 @@
+	if err != nil {
+		return errors.Wrap(err, "unable to retrieve configuration information")
+	}
+	do.EnvSpecificInfo = envInfo
+	do.Context = genericclioptions.NewDevfileContext(cmd)
+
+	return nil
+}
+
+// Validate validates the push parameters
+func (do *DeployOptions) Validate() (err error) {
+
+	log.Infof("\nValidation")
+
+	// Validate the --tag
+	s := log.Spinner("Validating arguments")
+
+	// Empty tag will be set in the adapter
+	// based on the namespace and component name
+	if do.tag != "" {
+		err = util.ValidateTag(do.tag)
+		if err != nil {
+			s.End(false)
+			return err
+		}
+	}
+	s.End(true)
+
+	do.devObj, err = devfile.ParseAndValidate(do.DevfilePath)
+	if err != nil {
+		return err
+	}
+
+	s = log.Spinner("Validating build information")
+	metadata := do.devObj.Data.GetMetadata()
+	dockerfileURL := metadata.Dockerfile
+
+	//Download Dockerfile to .odo, build, then delete from .odo dir
+	//If Dockerfile is present in the project already, use that for the build
+	//If Dockerfile is present in the project and field is in devfile, build the one already in the project and warn the user.
+	if dockerfileURL != "" && util.CheckPathExists(filepath.Join(do.componentContext, "Dockerfile")) {
+		// TODO: make clearer more visible output
+		log.Warning("Dockerfile already exists in project directory and one is specified in Devfile.")
+		log.Warningf("Using Dockerfile specified in devfile from '%s'", dockerfileURL)
+	}
+
+	if dockerfileURL != "" {
+		dockerfileBytes, err := util.LoadFileIntoMemory(dockerfileURL)
+		if err != nil {
+			s.End(false)
+			return errors.New("unable to download Dockerfile from URL specified in devfile")
+		}
+		// If we successfully downloaded the Dockerfile into memory, store it in the DeployOptions
+		do.DockerfileBytes = dockerfileBytes
+
+		// Validate the file that was downloaded is a Dockerfile
+		err = util.ValidateDockerfile(dockerfileBytes)
+		if err != nil {
+			s.End(false)
+			return err
+		}
+
+	} else if !util.CheckPathExists(filepath.Join(do.componentContext, "Dockerfile")) {
+		s.End(false)
+		return errors.New("dockerfile required for build. No 'alpha.build-dockerfile' field found in devfile, or Dockerfile found in project directory")
+	}
+
+	s.End(true)
+
+	s = log.Spinner("Validating deployment information")
+	manifestURL := metadata.Manifest
+
+	if manifestURL == "" {
+		s.End(false)
+		return errors.New("Unable to deploy as alpha.deployment-manifest is not defined in devfile.yaml")
+	}
+
+	manifestBytes, err := util.LoadFileIntoMemory(manifestURL)
+	if err != nil {
+		s.End(false)
+		return errors.Wrap(err, "unable to download manifest from URL specified in devfile")
+	}
+	do.ManifestSource = manifestBytes
+
+	// check if manifestSource contains {{.PORT}} template variable
+	// if it does, then check we have an port setup in env.yaml
+	do.DeploymentPort = 0
+	if bytes.Contains(manifestBytes, []byte("{{.PORT}}")) {
+		deploymentPort, err := do.EnvSpecificInfo.GetPortByURLKind(envinfo.ROUTE)
```

Comment:
I think this maybe should be revisited.    We'd like to enable the use case where, on OCP, you can push without doing `odo url create` , which means there will only be a devfile entry, but no env.yaml entry for the Url.